util/reggen/validate.py

```diff
@@ -662,21 +664,67 @@ def resolve_value(entry, param_list):
     return int(val), err
 
 
-# only supports subtractions right now
-def resolve_bitfield(entry, param_list):
-    lead_term = True
+# resolve expression
+def resolve_expression(entry, ops, param_list):
+    log.info("Evaluating entry {}".format(entry))
+
     val = 0
     error = 0
-    terms = entry.replace(" ", "").split('-')
 
-    for i in terms:
-        if lead_term:
-            lead_term = False
-            val, err = resolve_value(i, param_list)
+    # construct the pattern to search
+    operators = ''
+    for op in ops.keys():
+        operators += str(op)
+
+    pattern = '([{}])'.format(operators)
+
+    terms = re.split(pattern, entry)
+    log.info("Separate terms: {}".format(terms))
+
+    pending_op = ''
+
+    # When an operation is encountered, save its operation in pending op.
+    # On the next item, it is assumed to be a value and then the operation is
+    # performed.
+    for term in terms:
+        log.info("Operating on {}".format(term))
+        if term in ops.keys():
+            pending_op = term
         else:
-            tmp, err = resolve_value(i, param_list)
-            val -= tmp
+            op_val, err = resolve_value(term, param_list)
+            if pending_op:
+                log.info("Pending operation")
+                val = ops[pending_op](val, op_val)
+                pending_op = ''
+            else:
+                log.info("No pending operation")
+                val = op_val
+
         error += err
+
+    log.info("Final resolved value is {}".format(val))
+    return int(val), error
+
+
+# Supports addition / subtraction in bitfield calculation
+def resolve_bitfield(entry, param_list):
+    log.info("Resolving bitfield {}".format(entry))
+    val = 0
+    error = 0
+    ops = {"+": operator.add, "-": operator.sub}
+    entry = entry.replace(" ", "")
+    expression = re.search('[+-]', entry)
+
+    # The field is an expression
+    if expression:
```

Comment:
this part isn't really necessary... i guess i could have called `resolve_expression` directly...

Reply:
yes, it looks like this can be simplified since `resolve_expression` loops through term by term, and calls `resolve_value` on the individual terms. hence, `resolve_expression` can probably be arranged to handle a degenerate input correctly as well.

Reply:
but it feels weird to have resolve_bitfield do almost nothing...other than call resolve_expression hehe.
I guess it's okay, at least the abstraction is clear..and if we need to do more in the future this would be the place to add it.

Reply:
yeah, or we could also just remove `resolve_bitfield` if it ends up being an empty wrapper function...

Reply:
i actually just remembered why i might have added it.
The split function i'm using from re. returns nothing at all if there are
no operators.